File: backend/app/api/v1/endpoints/compliance.py
"""
Compliance & Data Protection API Endpoints

REST API for GDPR compliance, consent management, data subject requests,
retention policies, and privacy controls.
"""

from fastapi import APIRouter, HTTPException, Query
from typing import Optional

from app.schemas.compliance import (
    ConsentType, ConsentStatus, ConsentRecord, ConsentRequest, ConsentBulkRequest,
    ConsentPreferences, ConsentAuditLog, ConsentListResponse,
    DataSubjectRequest, DataSubjectRequestCreate, DataSubjectRequestUpdate,
    DataSubjectRequestType, DataSubjectRequestStatus, DataSubjectRequestListResponse,
    RetentionPolicy, RetentionPolicyCreate, RetentionPolicyUpdate,
    RetentionExecution, RetentionPolicyListResponse,
    ProcessingActivity, ProcessingActivityCreate, ProcessingActivityListResponse,
    EncryptionConfig, EncryptionConfigUpdate, EncryptionKey,
    PrivacyDocument, PrivacyDocumentCreate, PrivacyDocumentUpdate, PrivacyDocumentListResponse,
    AnonymizationRule, AnonymizationRuleCreate,
    DataBreach, DataBreachCreate, DataBreachUpdate, DataBreachListResponse,
    ComplianceStatus, ComplianceAuditListResponse,
)
from app.services.compliance_service import compliance_service

router = APIRouter()


# Consent Management

@router.get("/users/{user_id}/consent", response_model=ConsentPreferences, tags=["consent"])
async def get_consent_preferences(user_id: str):
    """Get user's consent preferences."""
    return compliance_service.get_consent_preferences(user_id)


@router.post("/users/{user_id}/consent", response_model=ConsentRecord, tags=["consent"])
async def update_consent(
    user_id: str,
    data: ConsentRequest,
    changed_by: Optional[str] = Query(None, description="User who made the change"),
):
    """Update user consent."""
    return compliance_service.update_consent(user_id, data, changed_by)


@router.post("/users/{user_id}/consent/bulk", tags=["consent"])
async def update_consent_bulk(user_id: str, data: ConsentBulkRequest):
    """Update multiple consents at once."""
    results = []
    for consent in data.consents:
        record = compliance_service.update_consent(user_id, consent)
        results.append(record)
    return {"records": results, "total": len(results)}


@router.post("/users/{user_id}/consent/withdraw-all", tags=["consent"])
async def withdraw_all_consent(
    user_id: str,
    ip_address: Optional[str] = Query(None),
):
    """Withdraw all non-essential consent for user."""
    count = compliance_service.withdraw_all_consent(user_id, ip_address)
    return {"success": True, "consents_withdrawn": count}


@router.get("/consent/records", response_model=ConsentListResponse, tags=["consent"])
async def list_consent_records(
    user_id: Optional[str] = Query(None),
    consent_type: Optional[ConsentType] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List consent records."""
    return compliance_service.list_consent_records(user_id, consent_type, skip, limit)


@router.get("/consent/audit", tags=["consent"])
async def get_consent_audit_logs(
    user_id: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """Get consent audit logs."""
    logs = compliance_service.get_consent_audit_logs(user_id, skip, limit)
    return {"logs": logs, "total": len(logs)}


# Data Subject Requests (GDPR)

@router.post("/dsr", response_model=DataSubjectRequest, tags=["dsr"])
async def create_data_subject_request(
    user_id: str = Query(...),
    data: DataSubjectRequestCreate = ...,
):
    """Create a GDPR data subject request."""
    return compliance_service.create_data_subject_request(user_id, data)


@router.get("/dsr", response_model=DataSubjectRequestListResponse, tags=["dsr"])
async def list_data_subject_requests(
    user_id: Optional[str] = Query(None),
    status: Optional[DataSubjectRequestStatus] = Query(None),
    request_type: Optional[DataSubjectRequestType] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List data subject requests."""
    return compliance_service.list_data_subject_requests(
        user_id, status, request_type, skip, limit
    )


@router.get("/dsr/overdue", tags=["dsr"])
async def get_overdue_requests():
    """Get all overdue data subject requests."""
    requests = compliance_service.get_overdue_requests()
    return {"requests": requests, "total": len(requests)}


@router.get("/dsr/{request_id}", response_model=DataSubjectRequest, tags=["dsr"])
async def get_data_subject_request(request_id: str):
    """Get data subject request by ID."""
    request = compliance_service.get_data_subject_request(request_id)
    if not request:
        raise HTTPException(status_code=404, detail="Request not found")
    return request


@router.patch("/dsr/{request_id}", response_model=DataSubjectRequest, tags=["dsr"])
async def update_data_subject_request(
    request_id: str,
    data: DataSubjectRequestUpdate,
    updated_by: str = Query(...),
):
    """Update data subject request."""
    request = compliance_service.update_data_subject_request(request_id, data, updated_by)
    if not request:
        raise HTTPException(status_code=404, detail="Request not found")
    return request


# Retention Policies

@router.post("/retention-policies", response_model=RetentionPolicy, tags=["retention"])
async def create_retention_policy(data: RetentionPolicyCreate):
    """Create a retention policy."""
    return compliance_service.create_retention_policy(data)


@router.get("/retention-policies", response_model=RetentionPolicyListResponse, tags=["retention"])
async def list_retention_policies(
    organization_id: Optional[str] = Query(None),
    data_category: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List retention policies."""
    return compliance_service.list_retention_policies(
        organization_id, data_category, skip, limit
    )


@router.get("/retention-policies/{policy_id}", response_model=RetentionPolicy, tags=["retention"])
async def get_retention_policy(policy_id: str):
    """Get retention policy by ID."""
    policy = compliance_service.get_retention_policy(policy_id)
    if not policy:
        raise HTTPException(status_code=404, detail="Policy not found")
    return policy


@router.patch("/retention-policies/{policy_id}", response_model=RetentionPolicy, tags=["retention"])
async def update_retention_policy(policy_id: str, data: RetentionPolicyUpdate):
    """Update retention policy."""
    policy = compliance_service.update_retention_policy(policy_id, data)
    if not policy:
        raise HTTPException(status_code=404, detail="Policy not found")
    return policy


@router.delete("/retention-policies/{policy_id}", tags=["retention"])
async def delete_retention_policy(policy_id: str):
    """Delete retention policy."""
    if compliance_service.delete_retention_policy(policy_id):
        return {"success": True, "message": "Policy deleted"}
    raise HTTPException(status_code=404, detail="Policy not found")


@router.post("/retention-policies/{policy_id}/execute", response_model=RetentionExecution, tags=["retention"])
async def execute_retention_policy(policy_id: str):
    """Execute a retention policy."""
    execution = compliance_service.execute_retention_policy(policy_id)
    if not execution:
        raise HTTPException(status_code=400, detail="Cannot execute policy (disabled, legal hold, or not found)")
    return execution


# Processing Activities (GDPR Article 30)

@router.post("/organizations/{org_id}/processing-activities", response_model=ProcessingActivity, tags=["processing"])
async def create_processing_activity(org_id: str, data: ProcessingActivityCreate):
    """Create processing activity record."""
    return compliance_service.create_processing_activity(org_id, data)


@router.get("/organizations/{org_id}/processing-activities", response_model=ProcessingActivityListResponse, tags=["processing"])
async def list_processing_activities(
    org_id: str,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List processing activities."""
    return compliance_service.list_processing_activities(org_id, skip, limit)


@router.get("/processing-activities/{activity_id}", response_model=ProcessingActivity, tags=["processing"])
async def get_processing_activity(activity_id: str):
    """Get processing activity by ID."""
    activity = compliance_service.get_processing_activity(activity_id)
    if not activity:
        raise HTTPException(status_code=404, detail="Activity not found")
    return activity


# Encryption

@router.get("/organizations/{org_id}/encryption", response_model=EncryptionConfig, tags=["encryption"])
async def get_encryption_config(org_id: str):
    """Get encryption configuration."""
    return compliance_service.get_encryption_config(org_id)


@router.patch("/organizations/{org_id}/encryption", response_model=EncryptionConfig, tags=["encryption"])
async def update_encryption_config(org_id: str, data: EncryptionConfigUpdate):
    """Update encryption configuration."""
    return compliance_service.update_encryption_config(org_id, data)


@router.post("/organizations/{org_id}/encryption/keys", response_model=EncryptionKey, tags=["encryption"])
async def create_encryption_key(
    org_id: str,
    name: str = Query(...),
    purpose: str = Query(...),
):
    """Create a new encryption key."""
    return compliance_service.create_encryption_key(org_id, name, purpose)


@router.get("/organizations/{org_id}/encryption/keys", tags=["encryption"])
async def list_encryption_keys(org_id: str):
    """List encryption keys."""
    keys = compliance_service.list_encryption_keys(org_id)
    return {"keys": keys, "total": len(keys)}


# Privacy Documents

@router.post("/privacy-documents", response_model=PrivacyDocument, tags=["privacy"])
async def create_privacy_document(data: PrivacyDocumentCreate):
    """Create privacy document."""
    return compliance_service.create_privacy_document(data)


@router.get("/privacy-documents", response_model=PrivacyDocumentListResponse, tags=["privacy"])
async def list_privacy_documents(
    document_type: Optional[str] = Query(None),
    organization_id: Optional[str] = Query(None),
    published_only: bool = Query(False),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List privacy documents."""
    return compliance_service.list_privacy_documents(
        document_type, organization_id, published_only, skip, limit
    )


@router.get("/privacy-documents/current", response_model=PrivacyDocument, tags=["privacy"])
async def get_current_privacy_document(
    document_type: str = Query(...),
    organization_id: Optional[str] = Query(None),
    locale: str = Query("en"),
):
    """Get current published privacy document."""
    document = compliance_service.get_current_privacy_document(
        document_type, organization_id, locale
    )
    if not document:
        raise HTTPException(status_code=404, detail="Document not found")
    return document


@router.get("/privacy-documents/{doc_id}", response_model=PrivacyDocument, tags=["privacy"])
async def get_privacy_document(doc_id: str):
    """Get privacy document by ID."""
    document = compliance_service.get_privacy_document(doc_id)
    if not document:
        raise HTTPException(status_code=404, detail="Document not found")
    return document


@router.post("/privacy-documents/{doc_id}/publish", response_model=PrivacyDocument, tags=["privacy"])
async def publish_privacy_document(doc_id: str):
    """Publish privacy document."""
    document = compliance_service.publish_privacy_document(doc_id)
    if not document:
        raise HTTPException(status_code=404, detail="Document not found")
    return document


# Anonymization

@router.post("/anonymization-rules", response_model=AnonymizationRule, tags=["anonymization"])
async def create_anonymization_rule(data: AnonymizationRuleCreate):
    """Create anonymization rule."""
    return compliance_service.create_anonymization_rule(data)


@router.get("/anonymization-rules", tags=["anonymization"])
async def list_anonymization_rules(
    organization_id: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List anonymization rules."""
    rules = compliance_service.list_anonymization_rules(organization_id, skip, limit)
    return {"rules": rules, "total": len(rules)}


# Data Breaches

@router.post("/organizations/{org_id}/breaches", response_model=DataBreach, tags=["breaches"])
async def report_data_breach(org_id: str, data: DataBreachCreate):
    """Report a data breach."""
    return compliance_service.report_data_breach(org_id, data)


@router.get("/organizations/{org_id}/breaches", response_model=DataBreachListResponse, tags=["breaches"])
async def list_data_breaches(
    org_id: str,
    resolved: Optional[bool] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List data breaches."""
    return compliance_service.list_data_breaches(org_id, resolved, skip, limit)


@router.get("/breaches/{breach_id}", response_model=DataBreach, tags=["breaches"])
async def get_data_breach(breach_id: str):
    """Get data breach by ID."""
    breach = compliance_service.get_data_breach(breach_id)
    if not breach:
        raise HTTPException(status_code=404, detail="Breach not found")
    return breach


@router.patch("/breaches/{breach_id}", response_model=DataBreach, tags=["breaches"])
async def update_data_breach(breach_id: str, data: DataBreachUpdate):
    """Update data breach."""
    breach = compliance_service.update_data_breach(breach_id, data)
    if not breach:
        raise HTTPException(status_code=404, detail="Breach not found")
    return breach


# Compliance Status

@router.get("/organizations/{org_id}/status", response_model=ComplianceStatus, tags=["compliance"])
async def get_compliance_status(org_id: str):
    """Get overall compliance status."""
    return compliance_service.get_compliance_status(org_id)


@router.get("/audit-logs", response_model=ComplianceAuditListResponse, tags=["compliance"])
async def list_compliance_audit_logs(
    resource_type: Optional[str] = Query(None),
    user_id: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List compliance audit logs."""
    return compliance_service.list_compliance_audit_logs(
        resource_type, user_id, skip, limit
    )


# User Data (GDPR Rights)

@router.get("/users/{user_id}/export", tags=["user-data"])
async def export_user_data(user_id: str):
    """Export all user data (GDPR portability)."""
    return compliance_service.export_user_data(user_id)


@router.delete("/users/{user_id}/data", tags=["user-data"])
async def delete_user_data(user_id: str):
    """Delete all user data (GDPR erasure / right to be forgotten)."""
    result = compliance_service.delete_user_data(user_id)
    return {"success": True, "deleted": result}

File: frontend/src/types/compliance.ts
/**
 * Compliance & Data Protection Types
 *
 * TypeScript types for GDPR compliance, consent management,
 * data subject requests, retention policies, and privacy controls.
 */

// Enums

export enum ConsentType {
  MARKETING = 'marketing',
  ANALYTICS = 'analytics',
  THIRD_PARTY = 'third_party',
  ESSENTIAL = 'essential',
  PERSONALIZATION = 'personalization',
  DATA_PROCESSING = 'data_processing',
  TERMS_OF_SERVICE = 'terms_of_service',
  PRIVACY_POLICY = 'privacy_policy',
}

export enum ConsentStatus {
  GRANTED = 'granted',
  DENIED = 'denied',
  WITHDRAWN = 'withdrawn',
  PENDING = 'pending',
  EXPIRED = 'expired',
}

export enum DataSubjectRequestType {
  ACCESS = 'access',
  RECTIFICATION = 'rectification',
  ERASURE = 'erasure',
  PORTABILITY = 'portability',
  RESTRICTION = 'restriction',
  OBJECTION = 'objection',
  AUTOMATED_DECISION = 'automated_decision',
}

export enum DataSubjectRequestStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  REJECTED = 'rejected',
  CANCELLED = 'cancelled',
}

export enum DataClassification {
  PUBLIC = 'public',
  INTERNAL = 'internal',
  CONFIDENTIAL = 'confidential',
  RESTRICTED = 'restricted',
  PII = 'pii',
  SENSITIVE = 'sensitive',
}

export enum RetentionAction {
  DELETE = 'delete',
  ANONYMIZE = 'anonymize',
  ARCHIVE = 'archive',
  REVIEW = 'review',
}

export enum LegalBasis {
  CONSENT = 'consent',
  CONTRACT = 'contract',
  LEGAL_OBLIGATION = 'legal_obligation',
  VITAL_INTERESTS = 'vital_interests',
  PUBLIC_INTEREST = 'public_interest',
  LEGITIMATE_INTERESTS = 'legitimate_interests',
}

export enum EncryptionAlgorithm {
  AES_256_GCM = 'aes-256-gcm',
  AES_256_CBC = 'aes-256-cbc',
  RSA_4096 = 'rsa-4096',
  CHACHA20_POLY1305 = 'chacha20-poly1305',
}

// Consent Interfaces

export interface ConsentRecord {
  id: string;
  user_id: string;
  consent_type: ConsentType;
  status: ConsentStatus;
  version: string;
  ip_address?: string;
  user_agent?: string;
  source: string;
  granted_at?: string;
  withdrawn_at?: string;
  expires_at?: string;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface ConsentRequest {
  consent_type: ConsentType;
  status: ConsentStatus;
  version: string;
  ip_address?: string;
  user_agent?: string;
  source?: string;
}

export interface ConsentBulkRequest {
  consents: ConsentRequest[];
}

export interface ConsentPreferences {
  user_id: string;
  marketing: ConsentStatus;
  analytics: ConsentStatus;
  third_party: ConsentStatus;
  personalization: ConsentStatus;
  essential: ConsentStatus;
  terms_accepted: boolean;
  terms_version?: string;
  privacy_accepted: boolean;
  privacy_version?: string;
  last_updated: string;
}

export interface ConsentAuditLog {
  id: string;
  user_id: string;
  consent_type: ConsentType;
  previous_status?: ConsentStatus;
  new_status: ConsentStatus;
  changed_by: string;
  reason?: string;
  ip_address?: string;
  created_at: string;
}

export interface ConsentListResponse {
  records: ConsentRecord[];
  total: number;
}

// Data Subject Request Interfaces

export interface DataSubjectRequest {
  id: string;
  user_id: string;
  request_type: DataSubjectRequestType;
  status: DataSubjectRequestStatus;
  email: string;
  description?: string;
  identity_verified: boolean;
  verification_method?: string;
  verified_at?: string;
  data_categories: string[];
  response_deadline: string;
  response_extended: boolean;
  extension_reason?: string;
  completed_at?: string;
  rejection_reason?: string;
  download_url?: string;
  download_expires_at?: string;
  notes: Array<Record<string, unknown>>;
  created_at: string;
  updated_at: string;
}

export interface DataSubjectRequestCreate {
  request_type: DataSubjectRequestType;
  email: string;
  description?: string;
  data_categories?: string[];
}

export interface DataSubjectRequestUpdate {
  status?: DataSubjectRequestStatus;
  identity_verified?: boolean;
  verification_method?: string;
  rejection_reason?: string;
  response_extended?: boolean;
  extension_reason?: string;
}

export interface DataSubjectRequestListResponse {
  requests: DataSubjectRequest[];
  total: number;
}

// Retention Policy Interfaces

export interface RetentionPolicy {
  id: string;
  name: string;
  description?: string;
  data_category: string;
  classification: DataClassification;
  retention_days: number;
  action: RetentionAction;
  legal_basis: LegalBasis;
  legal_hold: boolean;
  organization_id?: string;
  applies_to_tables: string[];
  applies_to_fields: string[];
  enabled: boolean;
  last_executed_at?: string;
  next_execution_at?: string;
  created_at: string;
  updated_at: string;
}

export interface RetentionPolicyCreate {
  name: string;
  description?: string;
  data_category: string;
  classification: DataClassification;
  retention_days: number;
  action: RetentionAction;
  legal_basis: LegalBasis;
  organization_id?: string;
  applies_to_tables?: string[];
  applies_to_fields?: string[];
  enabled?: boolean;
}

export interface RetentionPolicyUpdate {
  name?: string;
  description?: string;
  retention_days?: number;
  action?: RetentionAction;
  legal_basis?: LegalBasis;
  legal_hold?: boolean;
  applies_to_tables?: string[];
  applies_to_fields?: string[];
  enabled?: boolean;
}

export interface RetentionExecution {
  id: string;
  policy_id: string;
  policy_name: string;
  action: RetentionAction;
  records_affected: number;
  records_processed: number;
  records_failed: number;
  status: string;
  error_message?: string;
  started_at: string;
  completed_at?: string;
  duration_seconds?: number;
}

export interface RetentionPolicyListResponse {
  policies: RetentionPolicy[];
  total: number;
}

// Processing Activity Interfaces

export interface ProcessingActivity {
  id: string;
  name: string;
  description: string;
  purpose: string;
  legal_basis: LegalBasis;
  data_categories: string[];
  data_subjects: string[];
  recipients: string[];
  third_country_transfers: string[];
  transfer_safeguards?: string;
  retention_period: string;
  security_measures: string[];
  dpia_required: boolean;
  dpia_conducted: boolean;
  dpia_date?: string;
  controller_name: string;
  controller_contact: string;
  processor_name?: string;
  processor_contact?: string;
  dpo_contact?: string;
  organization_id: string;
  created_at: string;
  updated_at: string;
}

export interface ProcessingActivityCreate {
  name: string;
  description: string;
  purpose: string;
  legal_basis: LegalBasis;
  data_categories: string[];
  data_subjects: string[];
  recipients: string[];
  third_country_transfers?: string[];
  transfer_safeguards?: string;
  retention_period: string;
  security_measures?: string[];
  dpia_required?: boolean;
  controller_name: string;
  controller_contact: string;
  processor_name?: string;
  processor_contact?: string;
  dpo_contact?: string;
}

export interface ProcessingActivityListResponse {
  activities: ProcessingActivity[];
  total: number;
}

// Encryption Interfaces

export interface EncryptionKey {
  id: string;
  name: string;
  algorithm: EncryptionAlgorithm;
  key_size: number;
  purpose: string;
  status: string;
  version: number;
  created_at: string;
  rotated_at?: string;
  expires_at?: string;
  organization_id?: string;
}

export interface EncryptionConfig {
  organization_id: string;
  encrypt_at_rest: boolean;
  encrypt_in_transit: boolean;
  encrypt_pii: boolean;
  default_algorithm: EncryptionAlgorithm;
  key_rotation_days: number;
  auto_rotate_keys: boolean;
  pii_fields: string[];
  updated_at: string;
}

export interface EncryptionConfigUpdate {
  encrypt_at_rest?: boolean;
  encrypt_in_transit?: boolean;
  encrypt_pii?: boolean;
  default_algorithm?: EncryptionAlgorithm;
  key_rotation_days?: number;
  auto_rotate_keys?: boolean;
  pii_fields?: string[];
}

// Privacy Document Interfaces

export interface PrivacyDocument {
  id: string;
  document_type: string;
  version: string;
  title: string;
  content: string;
  effective_date: string;
  published: boolean;
  published_at?: string;
  organization_id?: string;
  locale: string;
  changelog?: string;
  requires_acceptance: boolean;
  created_at: string;
  updated_at: string;
}

export interface PrivacyDocumentCreate {
  document_type: string;
  version: string;
  title: string;
  content: string;
  effective_date: string;
  organization_id?: string;
  locale?: string;
  changelog?: string;
  requires_acceptance?: boolean;
}

export interface PrivacyDocumentUpdate {
  title?: string;
  content?: string;
  effective_date?: string;
  changelog?: string;
  requires_acceptance?: boolean;
}

export interface PrivacyDocumentListResponse {
  documents: PrivacyDocument[];
  total: number;
}

// Anonymization Interfaces

export interface AnonymizationRule {
  id: string;
  name: string;
  description?: string;
  field_name: string;
  table_name: string;
  technique: string;
  parameters: Record<string, unknown>;
  preserve_format: boolean;
  reversible: boolean;
  enabled: boolean;
  organization_id?: string;
  created_at: string;
}

export interface AnonymizationRuleCreate {
  name: string;
  description?: string;
  field_name: string;
  table_name: string;
  technique: string;
  parameters?: Record<string, unknown>;
  preserve_format?: boolean;
  reversible?: boolean;
  enabled?: boolean;
  organization_id?: string;
}

// Data Breach Interfaces

export interface DataBreach {
  id: string;
  title: string;
  description: string;
  severity: string;
  affected_records: number;
  affected_data_types: string[];
  affected_users: number;
  discovery_date: string;
  occurred_date?: string;
  contained_date?: string;
  resolved_date?: string;
  reported_to_authority: boolean;
  authority_report_date?: string;
  users_notified: boolean;
  notification_date?: string;
  root_cause?: string;
  remediation_steps: string[];
  organization_id: string;
  created_at: string;
  updated_at: string;
}

export interface DataBreachCreate {
  title: string;
  description: string;
  severity: string;
  affected_records: number;
  affected_data_types: string[];
  affected_users: number;
  discovery_date: string;
  occurred_date?: string;
}

export interface DataBreachUpdate {
  contained_date?: string;
  resolved_date?: string;
  reported_to_authority?: boolean;
  authority_report_date?: string;
  users_notified?: boolean;
  notification_date?: string;
  root_cause?: string;
  remediation_steps?: string[];
}

export interface DataBreachListResponse {
  breaches: DataBreach[];
  total: number;
}

// Compliance Status Interfaces

export interface ComplianceStatus {
  organization_id: string;
  gdpr_compliant: boolean;
  consent_coverage: number;
  pending_dsr_count: number;
  overdue_dsr_count: number;
  retention_policies_active: number;
  retention_policies_executed_today: number;
  processing_activities_count: number;
  data_breaches_open: number;
  encryption_enabled: boolean;
  last_audit_date?: string;
  next_audit_date?: string;
  compliance_score: number;
}

export interface ComplianceAuditLog {
  id: string;
  action: string;
  resource_type: string;
  resource_id: string;
  user_id: string;
  user_email: string;
  ip_address: string;
  details: Record<string, unknown>;
  created_at: string;
}

export interface ComplianceAuditListResponse {
  logs: ComplianceAuditLog[];
  total: number;
}

// Constants

export const GDPR_RESPONSE_DAYS = 30;
export const GDPR_EXTENSION_DAYS = 60;
export const BREACH_NOTIFICATION_HOURS = 72;

export const CONSENT_TYPE_LABELS: Record<ConsentType, string> = {
  [ConsentType.MARKETING]: 'Marketing Communications',
  [ConsentType.ANALYTICS]: 'Analytics & Statistics',
  [ConsentType.THIRD_PARTY]: 'Third-Party Sharing',
  [ConsentType.ESSENTIAL]: 'Essential Cookies',
  [ConsentType.PERSONALIZATION]: 'Personalization',
  [ConsentType.DATA_PROCESSING]: 'Data Processing',
  [ConsentType.TERMS_OF_SERVICE]: 'Terms of Service',
  [ConsentType.PRIVACY_POLICY]: 'Privacy Policy',
};

export const CONSENT_STATUS_LABELS: Record<ConsentStatus, string> = {
  [ConsentStatus.GRANTED]: 'Granted',
  [ConsentStatus.DENIED]: 'Denied',
  [ConsentStatus.WITHDRAWN]: 'Withdrawn',
  [ConsentStatus.PENDING]: 'Pending',
  [ConsentStatus.EXPIRED]: 'Expired',
};

export const DSR_TYPE_LABELS: Record<DataSubjectRequestType, string> = {
  [DataSubjectRequestType.ACCESS]: 'Right to Access',
  [DataSubjectRequestType.RECTIFICATION]: 'Right to Rectification',
  [DataSubjectRequestType.ERASURE]: 'Right to Erasure',
  [DataSubjectRequestType.PORTABILITY]: 'Right to Portability',
  [DataSubjectRequestType.RESTRICTION]: 'Right to Restriction',
  [DataSubjectRequestType.OBJECTION]: 'Right to Object',
  [DataSubjectRequestType.AUTOMATED_DECISION]: 'Automated Decision Making',
};

export const DSR_STATUS_LABELS: Record<DataSubjectRequestStatus, string> = {
  [DataSubjectRequestStatus.PENDING]: 'Pending',
  [DataSubjectRequestStatus.IN_PROGRESS]: 'In Progress',
  [DataSubjectRequestStatus.COMPLETED]: 'Completed',
  [DataSubjectRequestStatus.REJECTED]: 'Rejected',
  [DataSubjectRequestStatus.CANCELLED]: 'Cancelled',
};

export const DATA_CLASSIFICATION_LABELS: Record<DataClassification, string> = {
  [DataClassification.PUBLIC]: 'Public',
  [DataClassification.INTERNAL]: 'Internal',
  [DataClassification.CONFIDENTIAL]: 'Confidential',
  [DataClassification.RESTRICTED]: 'Restricted',
  [DataClassification.PII]: 'PII',
  [DataClassification.SENSITIVE]: 'Sensitive',
};

export const RETENTION_ACTION_LABELS: Record<RetentionAction, string> = {
  [RetentionAction.DELETE]: 'Delete',
  [RetentionAction.ANONYMIZE]: 'Anonymize',
  [RetentionAction.ARCHIVE]: 'Archive',
  [RetentionAction.REVIEW]: 'Review',
};

export const LEGAL_BASIS_LABELS: Record<LegalBasis, string> = {
  [LegalBasis.CONSENT]: 'Consent',
  [LegalBasis.CONTRACT]: 'Contract',
  [LegalBasis.LEGAL_OBLIGATION]: 'Legal Obligation',
  [LegalBasis.VITAL_INTERESTS]: 'Vital Interests',
  [LegalBasis.PUBLIC_INTEREST]: 'Public Interest',
  [LegalBasis.LEGITIMATE_INTERESTS]: 'Legitimate Interests',
};

// Helper Functions

export function getConsentStatusColor(status: ConsentStatus): string {
  const colors: Record<ConsentStatus, string> = {
    [ConsentStatus.GRANTED]: 'green',
    [ConsentStatus.DENIED]: 'red',
    [ConsentStatus.WITHDRAWN]: 'orange',
    [ConsentStatus.PENDING]: 'yellow',
    [ConsentStatus.EXPIRED]: 'gray',
  };
  return colors[status] || 'gray';
}

export function getDSRStatusColor(status: DataSubjectRequestStatus): string {
  const colors: Record<DataSubjectRequestStatus, string> = {
    [DataSubjectRequestStatus.PENDING]: 'yellow',
    [DataSubjectRequestStatus.IN_PROGRESS]: 'blue',
    [DataSubjectRequestStatus.COMPLETED]: 'green',
    [DataSubjectRequestStatus.REJECTED]: 'red',
    [DataSubjectRequestStatus.CANCELLED]: 'gray',
  };
  return colors[status] || 'gray';
}

export function getDataClassificationColor(classification: DataClassification): string {
  const colors: Record<DataClassification, string> = {
    [DataClassification.PUBLIC]: 'green',
    [DataClassification.INTERNAL]: 'blue',
    [DataClassification.CONFIDENTIAL]: 'yellow',
    [DataClassification.RESTRICTED]: 'orange',
    [DataClassification.PII]: 'red',
    [DataClassification.SENSITIVE]: 'purple',
  };
  return colors[classification] || 'gray';
}

export function getBreachSeverityColor(severity: string): string {
  const colors: Record<string, string> = {
    low: 'yellow',
    medium: 'orange',
    high: 'red',
    critical: 'purple',
  };
  return colors[severity] || 'gray';
}

export function calculateDSRDeadline(createdAt: string, extended: boolean = false): Date {
  const created = new Date(createdAt);
  const days = extended ? GDPR_EXTENSION_DAYS : GDPR_RESPONSE_DAYS;
  return new Date(created.getTime() + days * 24 * 60 * 60 * 1000);
}

export function isDSROverdue(request: DataSubjectRequest): boolean {
  if (
    request.status === DataSubjectRequestStatus.COMPLETED ||
    request.status === DataSubjectRequestStatus.REJECTED
  ) {
    return false;
  }
  return new Date() > new Date(request.response_deadline);
}

export function getDaysUntilDSRDeadline(request: DataSubjectRequest): number {
  const deadline = new Date(request.response_deadline);
  const now = new Date();
  const diffMs = deadline.getTime() - now.getTime();
  return Math.ceil(diffMs / (1000 * 60 * 60 * 24));
}

export function formatRetentionPeriod(days: number): string {
  if (days >= 365) {
    const years = Math.floor(days / 365);
    return years === 1 ? '1 year' : `${years} years`;
  }
  if (days >= 30) {
    const months = Math.floor(days / 30);
    return months === 1 ? '1 month' : `${months} months`;
  }
  return days === 1 ? '1 day' : `${days} days`;
}

export function formatComplianceScore(score: number): { label: string; color: string } {
  if (score >= 90) return { label: 'Excellent', color: 'green' };
  if (score >= 70) return { label: 'Good', color: 'blue' };
  if (score >= 50) return { label: 'Fair', color: 'yellow' };
  if (score >= 30) return { label: 'Poor', color: 'orange' };
  return { label: 'Critical', color: 'red' };
}

export function getConsentCoverageSummary(preferences: ConsentPreferences): string {
  const total = 4; // marketing, analytics, third_party, personalization
  let granted = 0;
  if (preferences.marketing === ConsentStatus.GRANTED) granted++;
  if (preferences.analytics === ConsentStatus.GRANTED) granted++;
  if (preferences.third_party === ConsentStatus.GRANTED) granted++;
  if (preferences.personalization === ConsentStatus.GRANTED) granted++;
  return `${granted} of ${total} consents granted`;
}

export function needsReConsent(
  currentVersion: string,
  acceptedVersion?: string
): boolean {
  if (!acceptedVersion) return true;
  // Simple version comparison - in production use semver
  return currentVersion !== acceptedVersion;
}
